engine/access/rest/blocks.go

```diff
@@ -194,31 +110,77 @@ func getBlockPayloadByID(req *requestDecorator, backend access.API, _ LinkGenera
 		return nil, NewBadRequestError(err.Error(), err)
 	}
 
-	_, payload, statusErr := blockLookup(req.Context(), id, backend)
-	if err != nil {
+	blkProvider := NewBlockProvider(backend, withID(&id))
+	blk, statusErr := blkProvider.getBlock(req.Context())
+	if statusErr != nil {
 		return nil, statusErr
 	}
-
+	payload := blockPayloadResponse(blk.Payload)
 	return payload, nil
 }
 
-func blockLookup(ctx context.Context, id flow.Identifier, backend access.API) (*generated.BlockHeader, *generated.BlockPayload, StatusError) {
-	flowBlock, err := backend.GetBlockByID(ctx, id)
-	if err != nil {
-		return nil, nil, idLookupError(id, "block", err)
+func getBlock(blkProvider *blockProvider, req *requestDecorator, backend access.API, link LinkGenerator) (*generated.Block, StatusError) {
+	var responseBlock = new(generated.Block)
+	responseBlock.Expandable = new(generated.BlockExpandable)
+
+	var id flow.Identifier
+	// if payload is to be expanded then lookup full block which contains both header and payload
+	if req.expands(ExpandableFieldPayload) {
+		blk, err := blkProvider.getBlock(req.Context())
+		if err != nil {
+			return nil, err
+		}
+		responseBlock.Header, responseBlock.Payload = blockHeaderResponse(blk.Header), blockPayloadResponse(blk.Payload)
+		id = blk.ID()
+	} else {
+
+		// else only lookup header and add expandable link for payload
+		header, statusErr := blkProvider.getHeader(req.Context())
+		if statusErr != nil {
+			return nil, statusErr
+		}
+		responseBlock.Header = blockHeaderResponse(header)
+		responseBlock.Payload = nil
+		id = header.ID()
+
+		payload, err := link.PayloadLink(id)
```

Comment:
I feel like this links can be moved to responses like here: https://github.com/onflow/flow-go/blob/cd5c42a7045ee8359de8e4db8fa6a3b42bc8db4b/engine/access/rest/converter.go#L326-L338 
Reasons for that is that anywhere that response is built the same functionality applies and we also have all the data, we can check if we have for example execution result and if we do then do nothing if we don't then generate the link. We can also just ignore this for now and then in the end think about how to maybe refactor this a bit, but to me this link generation doesn't yet feels right, because this handlers should really just be minimal an focus on getting resources, the links and how that is shaped has nothing to do with how to get resources. You can treat this for now just as an observation or a note and I would like to hear your opinion about it.